### ISSUE TITLE
Error on beta 28 - TypeError: undefined is not a function (evaluating 'addListener')

### ISSUE THREAD

**Issue by jslok** [react-navigation/react-navigation#3416]
### Current Behavior

When starting app on Android:
TypeError: undefined is not a function (evaluating 'addListener')

Not yet tested on iOS

### Expected Behavior

App should run

### How to reproduce

My app runs fine on react-navigation beta 27 and earlier. On beta 28 with no other changes, I get the error above. Here's a git repo which I created using ignite boilerplate and updated only react-navigation.
https://github.com/jslok/react-navigation-bug

npm install
react-native run-android

### Your Environment

| software         | version
| ---------------- | -------
| react-navigation | beta 28
| react-native     | 0.49.5 
| node             | 9.3
| npm or yarn      | npm

**Comment by Kennytian** [comment]
![image](https://user-images.githubusercontent.com/2621619/35766015-336d6482-0895-11e8-8050-c0af445daf62.png)

same issue +1

**Comment by Kennytian** [comment]
Maybe is here https://github.com/react-navigation/react-navigation/blob/189e7ee53906e6e2bf692f4b2f7a003d3f41ae52/src/views/CardStack/CardStack.js#L55

**Comment by joefazz** [comment]
You need to read the new Docs you need to add redux middleware and a new createReduxBoundAddListener property to the navigation helpers.

[Example](https://github.com/react-navigation/react-navigation/tree/master/examples/ReduxExample)

[Documentation](https://reactnavigation.org/docs/redux-integration.html)

**Comment by nico1510** [comment]
> but I'm not sure it's Github issue worthy
I'm sure it is ;-)

**Comment by brentvatne** [comment]
I updated the release notes to add this under breaking changes:

> * If you use Redux, please read https://reactnavigation.org/docs/redux-integration.html again!

**Comment by Dimous** [comment]
Will there be supplemental information about the nature of changes for alternative state containers (MobX in my case) users?

**Comment by jslok** [comment]
Thank you

**Comment by PatNeedham** [comment]
FYI anyone who is having a difficult time using the middleware included in the redux integration guide along with `redux-thunk` (or any other additional library), you can do something like this:

```
import ReduxThunk from 'redux-thunk';
import {
  createReduxBoundAddListener,
  createReactNavigationReduxMiddleware,
} from 'react-navigation-redux-helpers';


const NavMiddleware = createReactNavigationReduxMiddleware(
  "root",
  state => state.nav,
)

let middleware = [ReduxThunk, NavMiddleware]
const store = createStore(
  AppReducer,
  applyMiddleware(...middleware),
);

```

**Comment by kdenz** [comment]
We've the same need as Dimous's, how to do it with mobx?

**Comment by ngtlin** [comment]
Thx for the document update. It helped for me to fix the break.

**Comment by Laurensdc** [comment]
The docs are extremely confusing and not online with other Redux examples, especially for a first time Redux user... I can't wrap my head around it and have been stuck with this problem for days.

**Comment by jslok** [comment]
@laurensdc show your code. Someone may be able to help.

**Comment by denisschmidt** [comment]
who can help me? How to do it with mobx?

**Comment by Laurensdc** [comment]
@jslok Thanks, but I decided to just keep React Navigation separately.. Unfortunately I've already spent too much time on it, and I've got a bit of a tight deadline. I'll hopefully come back to it in a few weeks, appreciate the comment :)

**Comment by wkoutre** [comment]
@Laurensdc I think the docs -- as of today -- are pretty decent, actually, in terms of Redux support.

What would you change? Maybe we can put together a PR to add some clarification to make it easier for future users.

How to start debugging? Throw in some logs in the `node_modules/react-navigation/[whateverFile.js]`, reload, and trial/error your way through.

Anyway, I'm coming here because one of the apps I'm working on uses `beta21` and literally all of a sudden started crashing this morning because of Redux integration.

I have my store, Provider, Root navigator, etc -- all in different files, so had to carefully watch the sequence of importing from `index.js` and beyond. But @PatNeedham's code from the `react-navigation` docs did the trick for me.

Called `createReactNavigationReduxMiddleware` and added to my array of `middlewares` in my `configureStore.js` file.

In my `RootNavigation.js` -- the main Stack Navigator for this app -- I simply added:

```
  constructor(props) {
    super(props);

    this._addListener = createReduxBoundAddListener("root");
  }

...

render() {
    const { nav, dispatch } = this.props;

    return (
      <AppNavigator
        navigation={addNavigationHelpers({
          dispatch,
          state: nav,
          addListener: this._addListener
        })}
      />
    );
  }
```

This ensures `createReduxBoundAddListener("root");` is run after `createReactNavigationReduxMiddleware`.

Curious to know why this popped up all of the sudden this morning after months of developing the project with zero issues. But good to know this works.

**Comment by Laurensdc** [comment]
@wkoutre This was the same problem for me, I had my reducers, store, actions in separate files, and then the documentation for Redux implementation with React Navigation is a basic single file app.

For a simple integration (navigation state should be delegated by Redux), I've spent hours and hours of just trying to get the app not to crash. I had to just trial and error my way through.

Things that weren't (and still aren't) clear to me in the docs:
- The reducer does some magic which allows Redux to dispatch navigation actions with `const nextState = AppNavigator.router.getStateForAction(action, state);` - but the docs don't show how to actually dispatch a Navigation action (set the navigation state)
- `createReactNavigationReduxMiddleware`, `createReduxBoundAddListener` - what are they? What do they do? What is `root` in these functions?
- How to split up Navigator, reducers, store, etc.? Because some functions have to be called in the right order, splitting the code up can break the application.
- My need for letting Redux handle navigation state, was so I can navigate from within Redux. In other words, I'm using Thunk, which seems like a common case. However, the docs don't show this use case. 
- The docs use a very different solution than the `react-navigation/examples/ReduxExample` example, which uses a different solution from `react-navigation-redux-sample`. This makes everything more confusing than it should be.
- `AppNavigator.router.getStateForAction(AppNavigator.router.getActionForPathAndParams('Login'));` - Feels a bit verbose
- Now when it's implemented, how to navigate? Docs maybe should show some examples of navigation from within components and from within Redux actions.

These are just some of the confusions that I had while I was integrating Redux... Now I'm by no means an expert, but still this felt so damn unintuitive and difficult to implement (I somehow got it working now 👍 ).

**Comment by wkoutre** [comment]
I feel you @Laurensdc, and agree that the docs having everything in 1 file is a bit too “leightweight” for an example. But I feel it’s manageable with a decent understanding of Redux principles.

Could you explain more what you mean when you say your need for letting Redux handle navigation state was so you can navigate from within Redux? That’s like saying... I need water because I’m thirsty!

Navigation actions (all actions, really) are synchronous by default. `redux-thunk` is for adding `async` to your actions — aka, letting them take responsibility for, say, calls to the server to fetch new User info after a successful tap of the login button. You’re abstracting the logic from Components to make them as dumb and bug-free as possible. They know what they need to know, and nothing more. Ideally, they know nothing and just take in info and spit out UI.

I recommend [this video](https://egghead.io/courses/getting-started-with-redux) for understanding Redux core functionality under the hood so there’s no “magic” involved. It’s by Dan Abramov, the guy who took Flux principles and came up with the Redux API. There are 2 free ones by him that, if I recall correctly, have you build the fundamentals of Redux (connect, mapStateToProps, mapDispatchToProps, subscribers, verbose reducers) from the ground up.

In terms of what the helper functions do, #1-4 [in the docs’ intro](https://reactnavigation.org/docs/redux-integration.html) explain in 1-2 sentences. For a technical/functional understanding, just gotta dig into the code and/or play around with it.

Basically, react-navigation `dispatches` all navigation and handles the `navigation` state internally. Normally, this is not the same `dispatch` as the Redux `store`’s `dispatch`; however, by using the two helpers and passing in `dispatch` and `nav` (custom nav reducer created in your Redux `store`) to a `connect`ed `RootNavigator` component, you’re essentially:

- Replacing `react-navigation`’s `dispatch` with Redux’s `dispatch`
- Replacing the `state` of the `RootNavigator` to be the custom `nav` reducer you created

If you log out what `AppNavigator.router.getStateforAction(action, state)` is doing... it’s essentially saying:

- Is the action a `Navigation` action?
- If not, return the state
- If so, break down the action and find the next state for the `type` of `Navigation` action passed in

Types could be... NAVIGATE, RESET, BACK, etc.

How to navigate? `this.props.navigation.navigate(‘SomeScreen’)` is the most basic way. Integrated with Redux, it’s essentially just a wrapper around `dispatching` (using the Redux `store`’s `dispatch) a custom Navigation action.

```
const customNavAction = (screen) => ({
    {
        type: “Navigation/NAVIGATE”,
        screen,
    }
});

// All of these do the same thing
this.props.navigation.navigate(‘SomeScreen’);
this.props.navigation.dispatch(customNavAction(‘SomeScreen’);

// this assumes customNavAction is connected to the Component in mapDispatchToProps
this.props.customNavAction(‘SomeScreen’);
```

So, connected with Redux, you’re free to navigate from components that aren’t, necessarily, a declared `screen` in `RootNavigator` (or any other nav stack) — You can just dispatch an action.

You’ve created your own `nav` reducer in your store which is — by default via the boilerplate from the docs — responding to the `Navigation/<SOME_NAV_TYPE>` action `type`. But, you’re free to add in custom action `types`. For example, I have a `LOGOUT_SUCCESS` action `type` which triggers my `nav` reducer to reset my `RootNavigator` to my `Login` component.

Integrating with Redux, you can navigate **across navigators**, which can be very useful. So.. If my `HomeNav` is a `TabNavigator` with 5 tabs, and each tab is really a `StackNavigator`, handling its own state, and the app’s navigation ends up being like Spotify.

I just read through the Redux “full example ” and agree — confusing and not great code, either. Some of the boilerplate seems verbose, but in the end it’s a declarative wrapper around a relatively simplistic API.

Even with `AppNavigator.router.getStateForAction(AppNavigator.router.getActionForPathAndParams(‘Login’))`, it’s really saying:

- Take the AppNavigator
- Take its router (a `reducer`)
- Pass `Login` as the screen for a Navigation action
- Pass that `action` into the reducer `AppNavigator.router`
- Return the state determined by the reducer

TL:DR; Docs could be better, especially for Redux. If you know Redux, it’s relatively straightforward to piece it together. But it’s a chicken/egg problem if you don’t.

**Comment by brentvatne** [comment]
> TL:DR; Docs could be better, especially for Redux. If you know Redux, it’s relatively straightforward to piece it together. But it’s a chicken/egg problem if you don’t.

indeed. and if you don't know redux, this applies:

<img width="619" alt="screen shot 2018-03-06 at 2 37 09 pm" src="https://user-images.githubusercontent.com/90494/37062636-df2df808-214b-11e8-8651-b2696d125ceb.png">

**Comment by Laurensdc** [comment]
@wkoutre I'm doing nothing fancy or special. When a user logs in, the server replies. If it succeeded, redirect to another screen, otherwise show an error message.

I am not a Redux expert, it's the first time I'm using it. I very much see the appeal of it. 
It feels really wrong to handle all my state in Redux, but then set some variable `loggedIn` to true in the state, and in the `componentDidUpdate` check for `this.props.loggedIn` and then redirect with navigating to another screen. That's what I used to do and I don't like that solution, I feel like that whole logic (which screen should be active) should be tied with the state management, centrally in Redux.

I've watched the video tutorials and went through many examples. It's not magic, but it's not easy. I'm sure if you've used it in multiple projects before, it all makes perfect sense, but setting this whole thing up for a first time (which everyone has to do at one point) is very confusing.
I've also noticed that React Navigation is changing a lot, for example I found a lot of examples integrating redux without the need to set `addListener`. Okay, the docs state this very briefly, but still there's so much confusing information out there.

Thanks a lot for all the information and examples :) I feel like elaborate explanations like this are more than welcome in the documentation! This helps a lot.

**Comment by wkoutre** [comment]
@Laurensdc Again, I feel you on the first time setup and.. first time anything, really. I do things for the “first time” pretty much every day, lol.

In terms of storing navigation state in Redux, though... You can still do any/all logic you want while letting react-navigation maintain its own state. You’ll still have navigation state on every component that’s either:

- defined as a `screen` on a react-navigation `navigator`
- passed the `navigation` prop down from one of these screens ^^

It’ll be `this.props.navigation.state`. You can still navigate around by — one way, anyway — passing `navigation` to the async Redux action creator responsible for handling the login flow.

```
// LoginActions.js

import { loginToServer } from “../lib/LoginAPIHelpers” // or similar

const handleLogin = async ({ email, password, navigation }) => {
    return async dispatch => {
        dispatch({ type: “SET_LOADING” });
        
        Try {
                 const userPayload = await loginToServer({ email, password });

                 // do stuff with userPayload

                 dispatch({ type: LOGIN_SUCCESS, payload });
                 navigation.navigate(‘HomeScreen’)
        } catch(err) {
                const errorText = ... get error text from the error

                dispatch({ type: UNSET_LOADING });
                dispatch({ type: SET_LOGIN_ERROR, payload: errorText })
        }
    }
}
```

Even after having done a few projects with react-navigation integrated into Redux, I feel it’s just a matter of preference and doesn’t necessarily offer explicit functional benefits over letting nav keep its own state. Just my opinion, though.
